apps/snitch_core/test/domain/taxonomy/taxonomy_test.exs

```diff
@@ -0,0 +1,232 @@
+     ]}
+  end
+
+  describe "add_taxon/2" do
+    test "add child taxon" do
+      {_, [{target, _} | _]} = create_taxonomy()
+
+      child = %Taxon{name: "Carpet"}
+      taxon = Taxonomy.add_taxon(target, child, :child)
+
+      taxonomies = get_taxons_name_list(target)
+
+      assert taxon.name == "Carpet"
+      assert taxon.taxonomy_id == target.taxonomy_id
+      assert taxon.parent_id == target.id
+
+      assert taxonomies == [
+               "Home & Living",
+               "Flooring",
+               "Carpet",
+               "Kitchen & Tables",
+               "Table Covers",
+               "Mat & Napkins",
+               "Home Decor"
+             ]
+    end
+
+    test "add left taxon" do
+      {_, [{target, _} | _]} = create_taxonomy()
+
+      child = %Taxon{name: "Lamp and Lights"}
+      taxon = Taxonomy.add_taxon(target, child, :left)
+
+      taxonomies = get_taxons_name_list(target)
+
+      assert taxon.lft == 1
+      assert taxon.rgt == 2
+      assert taxon.parent_id == target.parent_id
+
+      assert taxonomies == [
+               "Home & Living",
+               "Lamp and Lights",
+               "Flooring",
+               "Kitchen & Tables",
+               "Table Covers",
+               "Mat & Napkins",
+               "Home Decor"
+             ]
+    end
+
+    test "add right taxon" do
+      {_, [{target, _} | _]} = create_taxonomy()
+
+      child = %Taxon{name: "Lamp and Lights"}
+      taxon = Taxonomy.add_taxon(target, child, :right)
+
+      taxons = get_taxons_name_list(target)
+
+      assert taxon.lft == 3
+      assert taxon.rgt == 4
+      assert taxon.parent_id == target.parent_id
+
+      assert taxons == [
+               "Home & Living",
+               "Flooring",
+               "Lamp and Lights",
+               "Kitchen & Tables",
+               "Table Covers",
+               "Mat & Napkins",
+               "Home Decor"
+             ]
+    end
+
+    test "add parent taxon" do
+      {_, [{target, _} | _]} = create_taxonomy()
+
+      child = %Taxon{name: "Lamp and Lights"}
+      taxon = Taxonomy.add_taxon(target, child, :parent)
+
+      taxons = get_taxons_name_list(target)
+
+      assert taxon.lft == 1
+      assert taxon.rgt == 4
+
+      assert taxons == [
+               "Home & Living",
+               "Lamp and Lights",
+               "Flooring",
+               "Kitchen & Tables",
+               "Table Covers",
+               "Mat & Napkins",
+               "Home Decor"
+             ]
+    end
+  end
+
+  describe "add_root/1" do
+    test "create root for empty taxonomy" do
+      taxonomy = insert(:taxonomy, name: "Home & Living")
+      taxon = %Taxon{name: "root", taxonomy_id: taxonomy.id}
+
+      root = Taxonomy.add_root(taxon)
+
+      assert root.name == taxon.name
+      assert root.taxonomy_id == taxonomy.id
+    end
+
+    test "create root for invalid taxonomy" do
+      taxon = %Taxon{name: "root", taxonomy_id: Faker.Number.digits()}
```

Comment:
This seems like the only use of Faker in the whole PR. I would have just put `-1` to get the same behaviour :smile: 
That said, this test is pointless, as it tests Ecto's behaviour.